File: emulinker/src/main/java/org/emulinker/kaillera/model/KailleraGame.java
package org.emulinker.kaillera.model;

import java.util.Collection;
import org.emulinker.kaillera.model.exception.*;
import org.emulinker.kaillera.model.impl.PlayerActionQueue;

public interface KailleraGame {
  public static final byte STATUS_WAITING = 0;
  public static final byte STATUS_PLAYING = 2;
  public static final byte STATUS_SYNCHRONIZING = 1;

  public static final String[] STATUS_NAMES = {"Waiting", "Playing", "Synchronizing"};

  public int getID();

  public String getRomName();

  public String getClientType();

  public PlayerActionQueue[] getPlayerActionQueue();

  public void setStartTimeout(boolean startTimeout);

  public boolean getStartTimeout();

  public void setSameDelay(boolean sameDelay);

  public boolean getSameDelay();

  public int getHighestPing();

  public int getDelay();

  public long getStartTimeoutTime();

  public KailleraUser getOwner();

  public void setDelay(int delay);

  public void setMaxUsers(int maxUsers);

  public int getMaxUsers();

  public int getStartN();

  public void setStartN(int startN);

  public boolean getP2P();

  public void setP2P(boolean p2P);

  public void setMaxPing(int maxPing);

  public int getMaxPing();

  public int getPlayerNumber(KailleraUser user);

  public int getNumPlayers();

  public KailleraUser getPlayer(int playerNumber);

  public Collection<? extends KailleraUser> getPlayers();

  public int getStatus();

  public KailleraServer getServer();

  public void droppedPacket(KailleraUser user);

  public int join(KailleraUser user) throws JoinGameException;

  public void chat(KailleraUser user, String message) throws GameChatException;

  public void kick(KailleraUser requester, int userID) throws GameKickException;

  public void start(KailleraUser user) throws StartGameException;

  public void ready(KailleraUser user, int playerNumber) throws UserReadyException;

  public void addData(KailleraUser user, int playerNumber, byte[] data) throws GameDataException;

  public void drop(KailleraUser user, int playerNumber) throws DropGameException;

  public void quit(KailleraUser user, int playerNumber)
      throws DropGameException, QuitGameException, CloseGameException;
}

File: emulinker/src/main/java/org/emulinker/kaillera/model/KailleraGame.kt
package org.emulinker.kaillera.model

import kotlin.Throws
import org.emulinker.kaillera.model.exception.CloseGameException
import org.emulinker.kaillera.model.exception.DropGameException
import org.emulinker.kaillera.model.exception.GameChatException
import org.emulinker.kaillera.model.exception.GameDataException
import org.emulinker.kaillera.model.exception.GameKickException
import org.emulinker.kaillera.model.exception.JoinGameException
import org.emulinker.kaillera.model.exception.QuitGameException
import org.emulinker.kaillera.model.exception.StartGameException
import org.emulinker.kaillera.model.exception.UserReadyException
import org.emulinker.kaillera.model.impl.PlayerActionQueue

interface KailleraGame {
  val clientType: String?
  val id: Int
  val owner: KailleraUser
  val playerActionQueue: Array<PlayerActionQueue>?
  val players: MutableList<KailleraUser>
  val romName: String
  val server: KailleraServer?
  val startTimeoutTime: Long
  val status: GameStatus

  var highestUserFrameDelay: Int
  var maxPing: Int
  var maxUsers: Int
  var ignoringUnnecessaryServerActivity: Boolean
  /** Frame delay is synced with others users in the same game (see /samedelay). */
  var sameDelay: Boolean
  var startN: Int
  var startTimeout: Boolean

  fun getPlayerNumber(user: KailleraUser): Int

  fun getPlayer(playerNumber: Int): KailleraUser?

  fun droppedPacket(user: KailleraUser)

  @Throws(JoinGameException::class)
  suspend fun join(user: KailleraUser): Int

  @Throws(GameChatException::class)
  fun chat(user: KailleraUser, message: String)

  @Throws(GameKickException::class)
  fun kick(requester: KailleraUser, userID: Int)

  @Throws(StartGameException::class)
  fun start(user: KailleraUser)

  @Throws(UserReadyException::class)
  fun ready(user: KailleraUser?, playerNumber: Int)

  @Throws(GameDataException::class)
  fun addData(user: KailleraUser, playerNumber: Int, data: ByteArray)

  @Throws(DropGameException::class)
  fun drop(user: KailleraUser, playerNumber: Int)

  @Throws(DropGameException::class, QuitGameException::class, CloseGameException::class)
  fun quit(user: KailleraUser, playerNumber: Int)
}
